doc/contributing/unicode.rst

```diff
@@ -0,0 +1,128 @@
+================
+Unicode handling
+================
+This document explains how Unicode and related issues are handled in CKAN.
+For a general introduction to Unicode and Unicode handling in Python 2 please
+read the `Python 2 Unicode HOWTO`_. Since Unicode handling differs greatly
+between Python 2 and Python 3 you might also be interested in the
+`Python 3 Unicode HOWTO`_.
+
+.. _Python 2 Unicode HOWTO: https://docs.python.org/2/howto/unicode.html
+.. _Python 3 Unicode HOWTO: https://docs.python.org/3/howto/unicode.html
+
+.. note::
+
+    This document describes the intended future state of Unicode handling in
+    CKAN. For historic reasons, some existing code does not yet follow the
+    rules described here.
+
+    *New code should always comply with the rules in this document. Exceptions
+    must be documented.*
+
+
+Overall Strategy
+----------------
+CKAN only uses Unicode internally (``unicode`` on Python 2). Conversion to/from
+ASCII strings happens on the boundary to other systems/libaries if necessary.
+
+
+Encoding of Python files
+------------------------
+Files containing Python source code (``*.py``) must be encoded using UTF-8, and
+the encoding must be declared using the following header::
+
+    # encoding: utf-8
+
+This line must be the first or second line in the file. See `PEP 263`_ for
+details.
+
+.. _PEP 263: https://www.python.org/dev/peps/pep-0263/
+
+
+String literals
+---------------
+String literals are string values given directly in the source code (as opposed
+to strings variables read from a file, received via argument, etc.). In
+Python 2, string literals by default have type ``str``. This has the risk of
+accidentially introducing ASCII strings and makes a future support of Python 3
+difficult (since string literals are interpreted as Unicode in Python 3).
+
+Therefore, the ``unicode_literals`` feature is used to make Python 2 treat
+string literals as Unicode::
+
+    from __future__ import unicode_literals
```

Comment:
There are many arguments for and against using `unicode_literals`, and the discussion in #3006 didn't result in a clear decision. This part is therefore my suggestion of how to deal with the issue, but of course I'm open to other opinions. Basically the two options are

1. Import `unicode_literals` and mark ASCII literals using the `b` suffix. This does the right thing most of the time. On the other hand, moving code between modules might break things, and you might end up with Unicode strings where you don't want to have them (e.g. filenames).
2. Don't import `unicode_literals` and use the `u` prefix for Unicode literals. Code can then be moved easily (because the meaning of a string literal only depends on its local definition) and you don't get Unicode strings by accident. However, you might get ASCII strings by accident (which I think is more problematic if you're trying to Unicode wherever it's possible). Since most string literals in the future should probably be Unicode, string literals without a prefix would be suspicious (intended ASCII literal or forgotten `u` prefix?), so they would either need a comment or (preferrably) a `b` prefix. This means that every single string literal in the source (including docstrings!) needs a `u` or a `b` prefix.

As I wrote in #3006, it's about choosing the lesser of two evils.